test/Microsoft.AspNetCore.Routing.Tests/DispatcherMiddlewareTest.cs

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing.Matchers;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Testing;
+using Microsoft.Extensions.Options;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Routing
+{
+    public class DispatcherMiddlewareTest
+    {
+        [Fact]
+        public async void Invoke_LogsCorrectValues_WhenNotHandled()
+        {
+            // Arrange
+            var expectedMessage = "Request did not match any endpoints.";
+
+            var sink = new TestSink(
+                TestSink.EnableWithTypeName<DispatcherMiddleware>,
+                TestSink.EnableWithTypeName<DispatcherMiddleware>);
+            var loggerFactory = new TestLoggerFactory(sink, enabled: true);
+
+            var httpContext = new DefaultHttpContext();
+            httpContext.RequestServices = new ServiceProvider();
+
+            RequestDelegate next = (c) => Task.FromResult<object>(null);
+
+            var logger = new Logger<DispatcherMiddleware>(loggerFactory);
+            var options = Options.Create(new DispatcherOptions());
+            var matcherFactory = new TestMatcherFactory(false);
+            var middleware = new DispatcherMiddleware(matcherFactory, options, logger, next);
+
+            // Act
+            await middleware.Invoke(httpContext);
+
+            // Assert
+            Assert.Empty(sink.Scopes);
+            var write = Assert.Single(sink.Writes);
+            Assert.Equal(expectedMessage, write.State?.ToString());
+        }
+
+        [Fact]
+        public async void Invoke_LogsCorrectValues_WhenHandled()
+        {
+            // Arrange
+            var expectedMessage = "Request matched endpoint 'Test endpoint'.";
+
+            var sink = new TestSink(
+                TestSink.EnableWithTypeName<DispatcherMiddleware>,
+                TestSink.EnableWithTypeName<DispatcherMiddleware>);
+            var loggerFactory = new TestLoggerFactory(sink, enabled: true);
+
+            var httpContext = new DefaultHttpContext();
+            httpContext.RequestServices = new ServiceProvider();
+
+            RequestDelegate next = (c) => Task.FromResult<object>(null);
+
+            var logger = new Logger<DispatcherMiddleware>(loggerFactory);
+            var options = Options.Create(new DispatcherOptions());
+            var matcherFactory = new TestMatcherFactory(true);
+            var middleware = new DispatcherMiddleware(matcherFactory, options, logger, next);
+
+            // Act
+            await middleware.Invoke(httpContext);
+
+            // Assert
+            Assert.Empty(sink.Scopes);
+            var write = Assert.Single(sink.Writes);
+            Assert.Equal(expectedMessage, write.State?.ToString());
+        }
+
```

Comment:
Is there any real functionality we should be testing here? 

I'll be honest, I'm on the record as strongly against adding tests for logging. We did them in a bunch of places in MVC and they have been nothing but a headache. In my experience from MVC it was a really bad tradeoff for us in terms of cost.

Reply:
Yeah tbh I'm not crazy about tests for logs except in certain situations. I'll refactor to test that underlying types are called.

Reply:
Set your ❤️ free then